app/src/main/kotlin/compiler/intermediate_form/UniqueIdentifier.kt

```diff
@@ -0,0 +1,72 @@
+package compiler.intermediate_form
+
+class IllegalCharacter(message: String) : RuntimeException(message)
+class InconsistentFunctionNamingConvention(message: String) : RuntimeException(message)
+
+class UniqueIdentifierFactory() {
+    companion object {
+        // it might be handy to change these values with ease
+        public val functionPrefix = "fun"
+        public val levelSeparator = '$'
+        public val polishSignSymbol = '#' // added after polish signs are converted to allowed characters
+        public val charactersAllowedByNasm = listOf('a'..'z', 'A'..'Z', '0'..'9').flatten() +
+            listOf('_', '$', '#', '@', '~', '.', '?')
+        public val knownConversionsToAllowedCharacters = mapOf(
+            'ą' to 'a',
+            'ć' to 'c',
+            'ę' to 'e',
+            'ł' to 'l',
+            'ó' to 'o',
+            'ś' to 's',
+            'ź' to 'z',
+            'ż' to 'z',
+            'Ą' to 'A',
+            'Ć' to 'C',
+            'Ę' to 'E',
+            'Ł' to 'L',
+            'Ó' to 'O',
+            'Ś' to 'S',
+            'Ź' to 'Z',
+            'Ż' to 'Z'
+        )
+        public val forbiddenLabels = listOf("globals")
+    }
+
+    private val knownIdentifiers: MutableSet<String> = mutableSetOf()
+    fun build(prefix: String?, curr: String): UniqueIdentifier {
+        fun convertToPlainASCII(character: Char): List<Char> {
+            if (character in charactersAllowedByNasm) return listOf(character)
+            if (character in knownConversionsToAllowedCharacters) return listOf(
+                knownConversionsToAllowedCharacters[character]!!,
+                polishSignSymbol // to make mangled names more readable
```

Comment:
One thing that bothers me is that we have two diacritic letters ('ż' and 'ź') translating into one ASCII char, so conflicts are possible.